src/lib/std-util/src/nested_ref.rs

```diff
@@ -0,0 +1,233 @@
+//! Tools for chaining borrows of [`std::cell::RefCell`].
+//!
+//! Useful for returning an abstract borrow, where multiple inner borrows are required. e.g.:
+//!
+//! ```
+//! use std::cell::RefCell;
+//!
+//! struct MyPrivateType {
+//!   x: RefCell<i32>
+//! }
+//!
+//! pub struct MyPublicType {
+//!   inner: RefCell<MyPrivateType>
+//! }
+//!
+//! impl MyPublicType {
+//!   pub fn borrow_x(&self) -> impl std::ops::Deref<Target=i32> + '_ {
+//!     std_util::nested_ref::NestedRef::map(self.inner.borrow(), |inner| inner.x.borrow())
+//!   }
+//! }
+//! ```
+//!
+//! Currently only supports nested RefCells; i.e. one level of nesting.
+//!
+//! TODO: It might be feasible to genericize the reference types, which would
+//! also add support for arbitrary levels of nesting.
+use std::cell::{Ref, RefMut};
+
+/// A nested [`std::cell::Ref`]. Useful for chaining borrows with [`std::cell::RefCell`].
+pub struct NestedRef<'a, Inner, Outer>
+where
+    Inner: 'static,
+{
+    // unsafely points to the Ref inside `_outer`.
+    // Must be declared first so that it's dropped first.
+    inner: Ref<'a, Inner>,
+    // Boxed so that `Self` is movable without breaking `inner`'s reference.
+    _outer: Box<Ref<'a, Outer>>,
+}
+impl<'a, Inner, Outer> NestedRef<'a, Inner, Outer>
+where
+    Inner: 'static,
+{
+    #[inline]
+    pub fn map(outer: Ref<'a, Outer>, borrow_fn: impl FnOnce(&Outer) -> Ref<Inner>) -> Self {
+        Self::filter_map(outer, |outer| Some(borrow_fn(outer))).unwrap()
+    }
+
+    #[inline]
+    pub fn filter_map(
+        outer: Ref<'a, Outer>,
+        borrow_fn: impl FnOnce(&Outer) -> Option<Ref<Inner>>,
+    ) -> Option<Self> {
+        let outer = Box::new(outer);
+        let inner: Ref<Inner> = borrow_fn(&outer)?;
+        let inner: Ref<'a, Inner> = unsafe { std::mem::transmute(inner) };
```

Comment:
> but it could do a lot of other things like return a borrow() of some other RefCell, or return a Ref that was created outside of the closure.

The compiler won't let the closure return something with a shorter lifetime than its input reference. It could perhaps return a `Ref<'static, Inner>` that doesn't borrow from `outer`, but that would still be safe.